### PR TITLE
Adding stl style reduce algorithm with tests.

### PR DESCRIPTION
**Pull request by andre-bergner**
I've add another missing STL algorithm to the algorithm header: `reduce` along with unit tests.

### CODE REVIEW

File: test/xsimd_algorithms.cpp [hunk 1]
```diff
@@ -8,102 +8,165 @@
...
 #include <iostream>
+#include <numeric>
 
 #include "xsimd/xsimd.hpp"
 
 #include "gtest/gtest.h"
 
 struct binary_functor
 {
-	template <class T>
```
[the remark this answers is not in the document]
Comment (JohanMabille) [reply test/xsimd_algorithms.cpp]: No problem, we are actually glad that you did it! We try to regularly fix the formatting but sometimes we miss things.

**Comment by JohanMabille** [comment]
It looks like it cannot donwload the intel framework for AVX512 emulation. I'm quite busy today, I'll have a look tomorrow (and review your PR at the same time)/

**Comment by wolfv** [comment]
cool!

**Comment by wolfv** [comment]
@andre-bergner are you planning other PR's? just in case -- there could be some overlap in development. Maybe you want to hop on gitter so we can discuss? https://gitter.im/QuantStack/Lobby :)

**Comment by andre-bergner** [comment]
Not right now. For following PRs I'll join on gitter ;)

**Pull request merged**